gitpod-io/gitpod
# server changes to enable new JB integration

## Description


It is an extract from https://github.com/gitpod-io/gitpod/pull/7362 with changes which should be deployed by @gitpod-io/engineering-meta team before.

There are following changes:
- add Gitpod plugin for JB Gateway as oauth2 client
- `getOwnerToken` method is added to allow Gitpod plugin to fetch it for SSH auth
- added a new referrer prefix `referrer:<client>(:<IDE>)?`, right now only `jetbrains-gateway` is supported, so one can start intellij directly by prefixing URL with `gitpod.io#referrer:jetbrains-gateway:intellij/`. Referrer prefix is also used to track clients access, so later can be reused for any derivative of VS Code for instance.
- add optional installation steps for desktop IDEs in order to open desktop links

## Related Issue(s)

Related to #7362, fixes #7525

## How to test


- Smoke test that you can start workspaces.
- You may try to use new referrer prefix `ak-jb-meta-changes.staging.gitpod-dev.com#referrer:jetbrains-gateway:intellij/` to check that it actually starts with Intellij. In segment for staging untrusted you can then find `ide-referrer` records.
  - a link to test: https://ak-jb-meta-changes.staging.gitpod-dev.com/#referrer:jetbrains-gateway:intellij/https://github.com/akosyakov/parcel-demo
    - It still should use CWM links, you should not see any installation instructions from us.
  - a link to test: https://ak-jb-meta-changes.staging.gitpod-dev.com/#referrer:vscode/https://github.com/akosyakov/parcel-demo
    - It should auto redirect and you should see installation instruction on workspace ready page.
<img width="1904" alt="Screenshot 2022-01-21 at 12 21 17" src="https://user-images.githubusercontent.com/3082655/150518648-2ea67596-8832-4179-9666-42f75c627f4c.png">

## Release Notes

```release-note
NONE
```

## Documentation


/werft analytics=segment|TEZnsG4QbLSxLfHfNieLYGF4cDwyFWoe
